src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SubSourceLogic.cs

```diff
@@ -450,191 +565,228 @@ public SubSourcePromiseControl(SourceShape<(
+        private readonly int _actorNumber;
+        private readonly Decider _decider;
 
-            public Outlet<TMessage> Out { get; }
-            public override SourceShape<TMessage> Shape { get; }
+        public Outlet<TMessage> Out { get; }
+        public override SourceShape<TMessage> Shape { get; }
 
-            public SubSourceStreamStage(TopicPartition topicPartition, IActorRef consumerActor,
-                                  Action<(TopicPartition, IControl)> subSourceStartedCallback,
-                                  Action<(TopicPartition, ISubSourceCancellationStrategy)> subSourceCancelledCallback,
-                                  IMessageBuilder<K, V, TMessage> messageBuilder,
-                                  Decider decider,
-                                  int actorNumber)
-            {
-                _topicPartition = topicPartition;
-                _consumerActor = consumerActor;
-                _subSourceStartedCallback = subSourceStartedCallback;
-                _subSourceCancelledCallback = subSourceCancelledCallback;
-                _messageBuilder = messageBuilder;
-                _decider = decider;
-                _actorNumber = actorNumber;
+        public SubSourceStage(TopicPartition topicPartition, IActorRef consumerActor,
+            Action<SubSourceStageLogicControl> subSourceStartedCallback,
+            Action<(TopicPartition, ISubSourceCancellationStrategy)> subSourceCancelledCallback,
+            Decider decider,
+            int actorNumber, ISubSourceStageLogicFactory<K, V, TMessage> subSourceStageLogicFactory)
+        {
+            _topicPartition = topicPartition;
+            _consumerActor = consumerActor;
+            _subSourceStartedCallback = subSourceStartedCallback;
+            _subSourceCancelledCallback = subSourceCancelledCallback;
+            _decider = decider;
+            _actorNumber = actorNumber;
+            _subSourceStageLogicFactory = subSourceStageLogicFactory;
+
+            Out = new Outlet<TMessage>("out");
+            Shape = new SourceShape<TMessage>(Out);
+        }
 
-                Out = new Outlet<TMessage>("out");
-                Shape = new SourceShape<TMessage>(Out);
-            }
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes)
+        {
+            return _subSourceStageLogicFactory.Create(Shape, _topicPartition, _consumerActor, 
+                _subSourceStartedCallback, _subSourceCancelledCallback, _actorNumber);
+        }
+    }
 
-            protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes)
+    /// <summary>
+    /// INTERNAL API
+    ///
+    /// A <see cref="SubSourceStageLogic{K,V,TMessage}"/> is the <see cref="GraphStageLogic"/>
+    /// of a <see cref="SubSourceStage{K,V,TMessage}"/>.
+    ///
+    /// This emits the actual Kafka messages downstream.
+    /// </summary>
+    internal abstract class SubSourceStageLogic<K, V, TMessage> : GraphStageLogic
```

Comment:
`SubSourceStageLogic` is where the real processing and interactivity with the `KafkaConsumerActor` happens. Most of the boilerplate is all wrapped inside this class but even still it should be subclassed for each flavor of stream stage it's going to be used with.